teensy4/usb_audio.cpp

```diff
@@ -323,114 +323,168 @@ static void copy_from_buffers(uint32_t *dst, int16_t *left, int16_t *right, unsi
-		if (right_2nd) { release(right_2nd); right_2nd = NULL; }
+	audio_block_t* chans[AUDIO_CHANNELS];
+	int i;
+	
+
+	// get the audio data
+	for (i=0;i<AUDIO_CHANNELS;i++)
+		chans[i] = receiveReadOnly(i);
+	
+	if (usb_audio_transmit_setting == 0) // not transmitting: dump all audio data
+	{
+		for (i=0;i<AUDIO_CHANNELS;i++)
+		{
+			if (NULL != chans[i]) 
+				release(chans[i]);
+			if (NULL != outgoing[i]) 
+			{
+				release(outgoing[i]);
+				outgoing[i] = NULL;
+			}
+				
+			if (NULL != ready[i]) 
+			{
+				release(ready[i]);
+				ready[i] = NULL;
+			}
+		}
 		offset_1st = 0;
-		return;
 	}
-	if (left == NULL) {
-		left = allocate();
-		if (left == NULL) {
-			if (right) release(right);
-			return;
+	else
+	{
+		// ensure every channel has a real audio block, even if it's silent
+		for (i=0;i<AUDIO_CHANNELS;i++)
+		{
+			if (NULL == chans[i]) // sent NULL: make implied silence into real data
+			{
+				chans[i] = allocate();
+				if (NULL != chans[i])
+					memset(chans[i]->data, 0, sizeof(chans[i]->data));
+				else
+					break; // no block available, exit early
+			}
 		}
-		memset(left->data, 0, sizeof(left->data));
-	}
-	if (right == NULL) {
-		right = allocate();
-		if (right == NULL) {
-			release(left);
-			return;
+		
+		if (i >= AUDIO_CHANNELS) // no invalid audio, queue all for transmission
+		{
+			__disable_irq();
+			
+			if (NULL == outgoing[0]) // just (re-)starting
+			{
+				for (i=0;i<AUDIO_CHANNELS;i++)
+					outgoing[i] = chans[i];
+				offset_1st = 0;
+			} 
+			else if (NULL == ready[0]) 
+			{
+				for (i=0;i<AUDIO_CHANNELS;i++)
+					ready[i] = chans[i];
+			} 
+			else 
+			{
+				// buffer overrun - PC is consuming too slowly
+				for (i=0;i<AUDIO_CHANNELS;i++)
+				{
+					audio_block_t* discard = outgoing[i];
+					outgoing[i] = ready[i];
+					ready[i] = chans[i];
+					release(discard);
+				}
+				offset_1st = 0; // TODO: discard part of this data?
+			}
+			__enable_irq();
+		}
+		else // some invalid audio, can't queue any - discard it all
+		{
+			for (i=0;i<AUDIO_CHANNELS;i++)
+				if (NULL != chans[i])
+					release(chans[i]);
+			
 		}
-		memset(right->data, 0, sizeof(right->data));
-	}
-	__disable_irq();
-	if (left_1st == NULL) {
-		left_1st = left;
-		right_1st = right;
-		offset_1st = 0;
-	} else if (left_2nd == NULL) {
-		left_2nd = left;
-		right_2nd = right;
-	} else {
-		// buffer overrun - PC is consuming too slowly
-		audio_block_t *discard1 = left_1st;
-		left_1st = left_2nd;
-		left_2nd = left;
-		audio_block_t *discard2 = right_1st;
-		right_1st = right_2nd;
-		right_2nd = right;
-		offset_1st = 0; // TODO: discard part of this data?
-		//serial_print("*");
-		release(discard1);
-		release(discard2);
 	}
-	__enable_irq();
+		
 }
 
 
+static void interleave_from_blocks(int16_t* transmit_buffer,	//!< next free sample in USB transmit buffer
+								audio_block_t** outgoing, //!< array of pointers to source audio blocks
+								int chans, //!< number of entries in the array
```

Comment:
nitpick: any reason for this to not just use the AUDIO_CHANNELS constant since it's a privately used function for this class?

Reply:
Only a vague thought that this interleaving code is fairly common in the library, and should at some point be pulled out into a super-optimised utility function, which would usually cater for several different "channel" counts.

Reply:
Fair enough! I don't mind too much either way :).